src/plugins/data/common/search/aggs/agg_types_registry.ts

```diff
@@ -60,11 +76,13 @@ export class AggTypesRegistry {
     };
 
     return {
-      get: (name: string) =>
-        getInitializedFromCache<BucketAggType<any> | MetricAggType<any>>(
-          name,
-          this.bucketAggs.get(name) || this.metricAggs.get(name)
-        ),
+      get: (name: string) => {
+        const agg =
+          this.bucketAggs.get(name) || this.metricAggs.get(name) || this.legacyAggs.get(name);
+        return agg
+          ? getInitializedFromCache<BucketAggType<any> | MetricAggType<any>>(name, agg)
+          : undefined;
```

Comment:
So, if I understand correctly, this is where the `undefined` possibility gets introduced that leads to the cascading changes in the tests and Lens data table.

At first I was concerned with this change to the `get` but, on further examination, I think it reflects reality. You could call `get` with a non-existing key and get an undefined value so the types should include that possibility.

However, on the consumer side I don't think I would ever expect to receive an undefined value from `get` as a valid runtime case. Since I probably have a strong dependency on whatever agg I'm using, I would probably either
- check for undefined and throw an error if the agg doesn't exist
- be tempted to simply append a `!` to avoid an extra check which I assume isn't necessary (as has been done in the tests)

All that is to say, I wonder if it makes more sense to guarantee a result from `get` by throwing an error in this code if the requested agg hasn't been registered. That simplifies things for the consumer's types and logic and "fails fast" if an app depends on a missing agg, making it easy to catch.

WDYT?

Reply:
> At first I was concerned with this change to the get but, on further examination, I think it reflects reality. You could call get with a non-existing key and get an undefined value so the types should include that possibility.

Right, the return type of `get` is already `AggType | undefined` but was not getting enforced by the compiler because of `getInitializedFromCache` parameter `agg` typed as `any`. I just added more specific typing that exposed this problem.

> However, on the consumer side I don't think I would ever expect to receive an undefined value from get as a valid runtime case.

Call consumers of `get` are already coded to handle `undefined` and the main use does already throw an exception when `get` returns undefined. See https://github.com/elastic/kibana/blob/main/src/plugins/data/common/search/aggs/agg_configs.ts#L194